### ISSUE TITLE
Failure to re-render after Line2D.set_color

### ISSUE THREAD

**Issue by Rufflewind** [matplotlib/matplotlib#7432]
> matplotlib.pyplot.draw()
> …
> This is used in interactive mode to update a figure that has been altered, but not automatically re-drawn. This should be only rarely needed, but there may be ways to modify the state of a figure with out marking it as stale. Please report these cases as bugs.

It appears that using `Line2D.set_color` does not cause matplotlib to re-render the figure.  This means calling `matplotlib.pyplot.draw()` or `fig.canvas.draw_idle()` is required when changing the line color on a plot interactively.

- [x] Matplotlib version, Python version and Platform (Windows, OSX, Linux ...)

    - Python version: 3.5.2
    - Matplotlib version: 1.5.3
    - Platform: Arch Linux 4.8.4 x86-64

- [x] How did you install Matplotlib and Python (pip, anaconda, from source ...)

    - Both via `pacman` (Arch Linux package manager)

- [x] If possible please supply a [Short, Self Contained, Correct, Example](http://sscce.org/) that demonstrates the issue i.e a small piece of code which reproduces the issue and can be run with out any other (or as few as possible) external dependencies.

    ~~~py
    import matplotlib
    import matplotlib.pyplot
    fig, [ax1, ax2] = matplotlib.pyplot.subplots(2)
    [line] = ax1.plot([0.0, 1.0], color="red")
    button = matplotlib.widgets.Button(ax2, "Greenify")
    def button_on_clicked(event):
        line.set_color("green")
        #fig.canvas.draw_idle()
    button.on_clicked(button_on_clicked)
    matplotlib.pyplot.show()
    ~~~

    1. Click the button. Avoid moving the mouse after clicking the button.
    2. Expected: The color should turn green.  Actuality: It stays red. (It can turn green if you move your mouse, however.)

- [ ] If this is an image generation bug attach a screenshot demonstrating the issue.
- [ ] If this is a regression (Used to work in an earlier version of Matplotlib), please 
      note where it used to work.

**Comment by tacaswell** [comment]
How are you running this and which GUI backend?

**Comment by Rufflewind** [comment]
I am running this via `python <script.py>` in graphical mode.  The GUI backend is `Qt5Agg`.

**Comment by tacaswell** [comment]
Ah, you are missing a `plt.ion()` call before you start plotting (which controls if the callbacks to trigger a re-draw when the figure in marked as stale are installed at figure creation time).

**Comment by Rufflewind** [comment]
Oh okay, I see.  Thanks for the clarification.

Perhaps the documentation could be clarified a bit?

~~~diff
diff --git a/lib/matplotlib/pyplot.py b/lib/matplotlib/pyplot.py
index 02180ee..b9782b7 100644
--- a/lib/matplotlib/pyplot.py
+++ b/lib/matplotlib/pyplot.py
@@ -674,10 +674,11 @@ def clf():
 def draw():
     """Redraw the current figure.
 
-    This is used in interactive mode to update a figure that has been
-    altered, but not automatically re-drawn.  This should be only rarely
-    needed, but there may be ways to modify the state of a figure with
-    out marking it as `stale`.  Please report these cases as bugs.
+    This is used to update a figure that has been altered, but not
+    automatically re-drawn.  If interactive mode is on (:func:`.ion()`), this
+    should be only rarely needed, but there may be ways to modify the state of
+    a figure without marking it as `stale`.  Please report these cases as
+    bugs.
 
     A more object-oriented alternative, given any
     :class:`~matplotlib.figure.Figure` instance, :attr:`fig`, that
~~~

**Comment by tacaswell** [comment]
@Rufflewind Could you open a PR with that change?

**Comment by Rufflewind** [comment]
#7449

**Comment by NelleV** [comment]
Fixed in #7449
